### PR TITLE
Support for Redmine 5.0.x, 4.2x and 4.1.x and Ruby up to 3.1.x

### PR DESCRIPTION
**Pull request by davidverdu**
In this PR we add support for the latest versions of Redmine. I have tested it in Redmine 5.0.1, 4.2.x and 4.1.0 with Ruby versions from 2.6.3 to 3.1.0

I also add the possibility of inheriting **Estimated Hours**

Before mergin this into master, I strongly recommend creating a branch **3.0.x** from the existing master to preserve the changes that allow running the plugin in lower versions of Redmine
```bash
git checkout master
git checkout -b 3.0.x
git push origin 3.0.x
```

### CODE REVIEW

File: app/views/subtasks_inherited_fields/_subtasks_inherited_fields.html.erb [hunk 1]
```diff
@@ -50,6 +50,11 @@
   <%= check_box_tag 'settings[inherit_is_private]', 1, @settings['inherit_is_private'] %>
 </p>
 
+<p>  
+  <label><%= l(:inherit_estimated_hours) %></label>
+  <%= check_box_tag 'settings[inherit_estimated_hours]', 1, @settings['inherit_estimated_hours'] %>
+</p>
+
```
Comment (davidverdu) [inline app/views/subtasks_inherited_fields/_subtasks_inherited_fields.html.erb:57]: Allow to config inheritance of  Estimated Hours

File: init.rb [hunk 2]
```diff
@@ -1,12 +1,16 @@
+require File.dirname(__FILE__) + '/lib/subtasks_inherited_fields.rb'
+require File.dirname(__FILE__) + '/lib/subtasks_inherited_fields/issues_helper_patch.rb'
+require File.dirname(__FILE__) + '/lib/subtasks_inherited_fields/issues_controller_patch.rb'
+
```
Comment (davidverdu) [inline init.rb:4]: New versions of Rails require defining the **require** this way

File: init.rb [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
   name 'Redmine Subtasks Inherited Fields plugin'
   author 'Edosoft Factory'
   description 'This is a plugin for Redmine to allow choosing which fields are inherited when you create a subtask'
-  version '1.1.1'
+  version '1.2.0'
   url 'http://www.edosoft.es'
   author_url 'mailto:david.verdu@edosoft.es'
   
-  requires_redmine version_or_higher: '3.0.0'
+  requires_redmine version_or_higher: '4.1.0'
```
Comment (davidverdu) [inline init.rb:13]: Minimum Redmine version is 4.1.0 but I recommend 5.0.x or 4.2.x

File: lib/subtasks_inherited_fields.rb [hunk 4]
```diff
@@ -24,6 +24,7 @@ def self.inherit_attrs(parent_issue)
       attrs[:description]      = parent_issue.description if settings['inherit_description']
       attrs[:is_private]       = parent_issue.is_private if settings['inherit_is_private']
       attrs[:status_id]        = parent_issue.status_id if settings['inherit_status_id']
+      attrs[:estimated_hours]  = parent_issue.estimated_hours if settings['inherit_estimated_hours']
```
Comment (davidverdu) [inline lib/subtasks_inherited_fields.rb:27]: inherit estimated hours

File: lib/subtasks_inherited_fields/issues_controller_patch.rb [hunk 5]
```diff
@@ -19,9 +18,11 @@ def redirect_after_create_plugin
           if params[:project_id]
             redirect_to new_project_issue_path(@issue.project, url_params)
           else
-            url_params[:issue].merge! :project_id => @issue.project_id
+            url_params[:issue][:project_id] = @issue.project_id
             redirect_to new_issue_path(url_params)
           end
+        elsif params[:follow]
+          redirect_to issue_path(@issue)
```
Comment (davidverdu) [inline lib/subtasks_inherited_fields/issues_controller_patch.rb:25]: Include changes for the recent Redmine

File: config/locales/en.yml [hunk 6]
```diff
@@ -12,4 +12,5 @@ en:
   inherit_description: "Inherit parent Description"
   inherit_is_private: "Inherit parent Private"
   inherit_parent_tracker: "Inherit parent Tracker"
+  inherit_estimated_hours: "Inherit parent Estimated Hours"
```
Comment (davidverdu) [inline config/locales/en.yml:15]: i18n for the new inherited field

**Comment by ashrafalzyoud** [comment]
perfect @davidverdu

File: init.rb [hunk 7]
```diff
@@ -1,12 +1,16 @@
+require File.dirname(__FILE__) + '/lib/subtasks_inherited_fields.rb'
+require File.dirname(__FILE__) + '/lib/subtasks_inherited_fields/issues_helper_patch.rb'
+require File.dirname(__FILE__) + '/lib/subtasks_inherited_fields/issues_controller_patch.rb'
+
 Redmine::Plugin.register :redmine_subtasks_inherited_fields do
   name 'Redmine Subtasks Inherited Fields plugin'
   author 'Edosoft Factory'
   description 'This is a plugin for Redmine to allow choosing which fields are inherited when you create a subtask'
-  version '1.1.1'
+  version '1.2.0'
```
Comment (davidverdu) [inline init.rb:9]: Increase plugin version

File: test_cases/subtasks_inherited_fields.feature [hunk 8]
```diff
@@ -46,6 +46,7 @@ Feature: Subtasks configurable inherited fields
     And I see Category has a value
     And I see Assignee has a value
     And I see Status has a value
+    And I see Estimated Hours has a value
```
Comment (davidverdu) [inline test_cases/subtasks_inherited_fields.feature:49]: add new inherited field to specs

**Comment by ashrafalzyoud** [comment]
merge it plz to download it
@davidverdu

**Comment by davidverdu** [comment]
> perfect man its working good thx for your support @davidverdu

You are welcome!

**Comment by ashrafalzyoud** [comment]
can u let it empty if im not choose
in my case i dont need traker _id when i create subtask
`Default subtask Tracker (when is not inherited from parent task)`

**Comment by davidverdu** [comment]
> can u let it empty if im not choose in my case i dont need traker _id when i create subtask `Default subtask Tracker (when is not inherited from parent task)` @davidverdu

Hey @ashrafalzyoud . As far as I know, in Redmine there is no such thing as an Issue without a Tracker associated to it. With the plugin you have two options: Inherit the parent tracker or define a default tracker for subtasks. For example, you can define parent issues as UserStory and then decide that subtasks have the tracker Task.

**Pull request merged**